core/store/models/run_result.go

```diff
@@ -0,0 +1,42 @@
+package models
+
+import (
+	"fmt"
+
+	"github.com/tidwall/gjson"
+	null "gopkg.in/guregu/null.v3"
+)
+
+// RunResult keeps track of the outcome of a TaskRun or JobRun. It stores the
+// Data and ErrorMessage, and contains a field to track the status.
+type RunResult struct {
+	ID           uint        `json:"-" gorm:"primary_key;auto_increment"`
+	Data         JSON        `json:"data" gorm:"type:text"`
+	ErrorMessage null.String `json:"error"`
+}
+
+// SetError marks the result as errored and saves the specified error message
+func (rr *RunResult) SetError(err error) {
+	if err != nil {
+		rr.ErrorMessage = null.StringFrom(err.Error())
+	}
+}
+
+// ResultString returns the string result of the Data JSON field.
+func (rr RunResult) ResultString() (string, error) {
+	val := rr.Result()
+	if val.Type != gjson.String {
+		return "", fmt.Errorf("non string result")
```

Comment:
Why are only strings allowed? Doesn't that defeat the purpose of using json, to some extent?

Reply:
This is another carry over. I think this is really just an optimisation for cases where only a string is allowed, to reduce the lines of code written in the caller. The interface also supports raw access to the Data.